### DISCUSSION TITLE
Default namespace not working as what I expected

### DISCUSSION THREAD

**Discussion by ngtrthinh169TPC** [i18next/i18next#2067] [Q&A] [answered]
I'm sorry that this is not really something contributing to the project. It's only my bad job at setting up but I can't find any helps elsewhere (posted something similar on StackOverflow but no answer). If this is not suitable, please tell me somewhere I can ask this, I'll be delightfully move this question there.

I'm trying to use i18next with create-react-app and set up similar to [the official docs of i18next](https://www.i18next.com/overview/typescript). But my intellisense and typescript doesn't look right.

My translation json `en.json`:

```json
{
  "example": "text"
}
```

`i18n.d.ts`:

```js
import "i18next";
import { resources } from "../i18n";

declare module "i18next" {
  interface CustomTypeOptions {
    resources: (typeof resources)["en"];
  }
}
```

`i18n.ts` (since I imported `resources` from here):
```js
import i18n from "i18next";
import { initReactI18next } from "react-i18next";

import en from "./locale/en.json";

export const resources = {
  en: {
    translation: en,
  },
};

i18n.use(initReactI18next).init({
  resources,
  lng: "en",
  interpolation: {
    escapeValue: false,
  },
});

export default i18n;
```

I tried to use the t function in my `App.tsx`:

```jsx
import "./App.css";
import { useTranslation } from "react-i18next";

const App = () => {
  const { t } = useTranslation();

  return (
    <div className="App">
      {/* This does not work */}
      <p>{t("example")}</p>
      {/* This does not work either */}
      <p>{t("translation.example")}</p>
      {/* This works, but unexpected */}
      <p>{t("translation:example")}</p>
    </div>
  );
};

export default App;
```

As far as I know, i18next has the default namespace of `translation`, so I should not need to provide that namespace on usage.
Actually I think it was ok to add that namespace, so I tried to add it to the t function:

```jsx
  const { t } = useTranslation("translation");
```

But it doesn't change anything of intellisense and typescript error.

For context, I'm using latest version for the time being: `i18next` v23.7.6, `react-i18next` v13.5.0, `typescript` v5.2.2

I want to know why and how this interaction work, since I'm working with some legacy code and cannot overhaul the whole thing (beforehand it doesn't use typescript and has an older version of i18next).

Edit: My minimal reproduction of what I'm doing: https://github.com/ngtrthinh169TPC/rep-i18next-ts . Without context of the legacy code, the interaction is still the same.

Edit 2: As a side note: my t function returns type `never`, so I really think that I'm doing the set up wrong somewhere.

**Comment by adrai** [comment]
https://github.com/ngtrthinh169TPC/rep-i18next-ts/pull/1/files

**Reply by ngtrthinh169TPC** [reply]
I'm not sure but doesn't seem that this fixes my problem. There's no `en` on `en.json`, so

```js
import "i18next";
// import { resources } from "../i18n"; // seems your exported resources are not the same as importing directly here
import resources from "./locale/en.json";

declare module "i18next" {
  interface CustomTypeOptions {
    resources: typeof resources.en;
  }
}
```

this fix of yours doesn't work :( it suppressed the error but just because of `typeof resources.en` returns `any`

**Reply by adrai** [reply]
Sorry, missed this change:

```diff
- resources: typeof resources.en;
+ resources: typeof resources;
```

**Reply by ngtrthinh169TPC** [reply]
Another side note: I tried using `i18next-resources-for-ts`, but no luck. I also tried that script and change resources type into `resources["en"]["translation"]` but somehow it changes `example` into a namespace.

My conclusion is I don't really understand how i18next's resources type work, so it's gonna be awesome if you can explain a bit about it.

**Reply by ngtrthinh169TPC** [reply]
My d.ts file is currently looking like this
```js
import "i18next";
import resources from "../locale/en.json";

declare module "i18next" {
  interface CustomTypeOptions {
    resources: typeof resources;
  }
}
```

And as I've said above, strangely the `example` key changed into a namespace
![image](https://github.com/i18next/i18next/assets/74962522/4b3868e5-f1c1-4b2b-9c2f-ad92c756a0ea)

**Reply by adrai** [reply]
Ahh now I know what you mean.
I suspect this is because of yarn.
Can you try to reproduce it in this example? With npm not yarn first...

https://github.com/i18next/react-i18next/tree/master/example/react-typescript/simple

**Reply by ngtrthinh169TPC** [reply]
Now it makes sense, I can't install `react-i18next` on npm because of `typescript` conflicted dependency with `react-scripts`

**Reply by adrai** [reply]
react-scripts seems to require ts 4.x

**Reply by adrai** [reply]
Can you downgrade ts to v4?

**Reply by ngtrthinh169TPC** [reply]
It seems that we'll have to find a way to work around with `react-scripts`.
I'd be thankful if you have any recommendations to solve this, but hey, it's javascript, can't expect everything to work harmonically.

I'll update as soon as I can find anything. Thank you for your time.

**Reply by ngtrthinh169TPC** [reply]
I'll try that one

**Reply by ngtrthinh169TPC** [reply]
Tried replicate your example above with typescript 4.9.5. Both npm and yarn worked.
Any conclusions from your side? Since I'm really curious about what's going on here

**Comment by adrai** [comment, marked as answer]
That's the problem:
<img width="367" alt="image" src="https://github.com/i18next/i18next/assets/1086194/3655eb44-0780-43ea-8c53-6442197b3a79">
<img width="772" alt="image" src="https://github.com/i18next/i18next/assets/1086194/c1cbb7c9-f36e-440c-9104-6c378a21fd5b">

**Reply by ngtrthinh169TPC** [reply]
Thank you very much. It solves my problem completely. I'll suggest marking this a warning rather than information, since it's important to not skip.
Or it was just empty-minded me.

**Reply by FloStar3000** [reply]
Thanks for the solution, but for me, enabling strictNullChecks or strict won't work for me, i get at least 900 compiler errors when i turn either of them on.

**Reply by FloStar3000** [reply]
Ok so i use Next.js and it's a bigger project that never had "strict" or "strictNullChecks" turned on. I think it's a good idea to turn on strict mode, but there is no way that i am fixing all the 2000 TypeScript compiler errors that result out of this at once. I developed a semi-sketchy workaround, so i18next works with TypeScript but i don't have to fix all the TS errors right away.
I set "strict" and "strictNullChecks" to true in my tsconfig.json, this will let me run my project in dev environment and also when i open any file, it highlights me the errors that result from turning on strict mode, so i can fix them on the go if i want to. 
For building in production, i simply _remove_ the type definitions for i18next, that way i can disable strict and strictNullChecks again and the TS compilation will succeed without any errors.
I moved my tsconfig.json to a `tsconfig-base.json`. 
Then i created a `tsconfig-dev.json`:
```
{
	"extends": "./tsconfig-base.json",
	"compilerOptions": {
		"strict": true,
		"strictNullChecks": true
	}
}
```
and a `tsconfig-build.json`
```
{
	"extends": "./tsconfig-base.json",
	"compilerOptions": {
		"strict": false,
		"strictNullChecks": false
	},
	"exclude": ["types/i18next.d.ts"]
}
```
I simply copy the `tsconfig-dev.json` to `tsconfig.json` when i'm developing and use the `tsconfig-build.json` when i build the project.
